### PR TITLE
fix adagrad error

### PR DESCRIPTION
**Pull request by chengxiaohua1105**
fix the bug of adagrad algorithm.

The adagrad algorithm is implemented as follows:

for (size_t i = 0; i < param.size(); ++i) {                        
    accum_g[i] += grad[i] * grad[i];                                 
    param[i] += learning_rate * grad[i] / std::sqrt(accum_g[i] + epsilon_)  + learning_rate * decay_ * param[i];                  
}

 it should be implemented as follows

for (size_t i = 0; i < param.size(); ++i) {                        
    accum_g[i] += grad[i] * grad[i];                                 
    param[i] -= learning_rate * grad[i] / std::sqrt(accum_g[i] + epsilon_) + learning_rate * decay_ * param[i];                  
}

### CODE REVIEW

**Comment by tonyyang-svail** [comment]
@abhinavarora please confirm this. It looks suspicious to me.

**Comment by abhinavarora** [comment]
The correction from `+=` to `-=` is definitely correct. However, the way the weight decay is being added is mathematically incorrect. We should be modifying the gradient first and then using the modified gradient for the updates. However, I checked that this implementation is incorrect in all optimizers. We can refer to http://pytorch.org/docs/0.1.12/optim.html for more details.
@chengxiaohua1105  Should we open an issue for all optimizers?

**Comment by luotao1** [comment]
close due to `paddle/optimizer/adagrad_optimizer.cc` are not in the develop branch now.
